File: src/main/swift/Sources/swift/fr/jhelp/kotlinLigthTools/atomic/AtomicReference.swift


class AtomicReference<T> {

    private let mutex: Mutex
    private var value: T?

    init(_ value: T?){
        self.mutex = Mutex()
        self.value = value
    }

    func get() -> T?
    {
        var value: T? = nil
        self.mutex.safeExecute { value = self.value }
        return value
    }

    func set(_ value: T?)
    {
        self.mutex.safeExecute { self.value = value }
    }

    func getAndSet(_ newValue: T?) -> T?
    {
        var value: T? = nil

        self.mutex.safeExecute {
            value = self.value
            self.value = newValue
        }

        return value
    }
}

File: src/main/kotlin/fr/jhelp/kotlinLigthTools/atomic/AtomicReference.kt
package fr.jhelp.kotlinLigthTools.atomic

import fr.jhelp.kotlinLight.Mutex

class AtomicReference<T>
{
    private val mutex: Mutex
    private var value: T?

    constructor(value: T?)
    {
        this.mutex = Mutex()
        this.value = value
    }

    fun get(): T?
    {
        var value: T? = null
        this.mutex.safeExecute { value = this.value }
        return value
    }

    fun set(value: T?)
    {
        this.mutex.safeExecute { this.value = value }
    }

    fun getAndSet(newValue: T?): T?
    {
        var value: T? = null

        this.mutex.safeExecute {
            value = this.value
            this.value = newValue
        }

        return value
    }
}